plugins/cache-caffeine/src/test/java/CaffeineHeapCacheTests.java

```diff
@@ -0,0 +1,451 @@
+        }
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            caffeineTest.put(entry.getKey(), entry.getValue());
+        }
+
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            threads[j] = new Thread(() -> {
+                phaser.arriveAndAwaitAdvance();
+                assertEquals(entry.getValue(), caffeineTest.get(entry.getKey()));
+                countDownLatch.countDown();
+            });
+            threads[j].start();
+            j++;
+        }
+        phaser.arriveAndAwaitAdvance(); // Will trigger parallel puts above.
+        countDownLatch.await(); // Wait for all threads to finish
+        assertEquals(randomKeys, caffeineTest.stats().getTotalHits());
+        caffeineTest.close();
+    }
+
+    public void testConcurrentPut() throws Exception {
+        ToLongBiFunction<ICacheKey<String>, String> weigher = getMockWeigher();
+        MockRemovalListener<String, String> removalListener = new MockRemovalListener<>();
+        ICache<String, String> caffeineTest = new CaffeineHeapCache.Builder<String, String>()
+            .setDimensionNames(List.of(dimensionName))
+            .setExecutor(Runnable::run)
+            .setExpireAfterAccess(TimeValue.MAX_VALUE)
+            .setMaximumWeightInBytes(CACHE_SIZE_IN_BYTES)
+            .setWeigher(weigher)
+            .setRemovalListener(removalListener)
+            .build();
+        int randomKeys = randomIntBetween(20, 100);
+        Thread[] threads = new Thread[randomKeys];
+        Phaser phaser = new Phaser(randomKeys + 1);
+        CountDownLatch countDownLatch = new CountDownLatch(randomKeys);
+        Map<ICacheKey<String>, String> keyValueMap = new HashMap<>();
+        int j = 0;
+        for (int i = 0; i < randomKeys; i++) {
+            keyValueMap.put(getICacheKey(UUID.randomUUID().toString()), UUID.randomUUID().toString());
+        }
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            threads[j] = new Thread(() -> {
+                phaser.arriveAndAwaitAdvance();
+                caffeineTest.put(entry.getKey(), entry.getValue());
+                countDownLatch.countDown();
+            });
+            threads[j].start();
+            j++;
+        }
+        phaser.arriveAndAwaitAdvance(); // Will trigger parallel puts above.
+        countDownLatch.await(); // Wait for all threads to finish
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            String value = caffeineTest.get(entry.getKey());
+            assertEquals(entry.getValue(), value);
+        }
+        assertEquals(randomKeys, caffeineTest.count());
+        assertEquals(randomKeys, caffeineTest.stats().getTotalItems());
+        caffeineTest.close();
+    }
+
+    public void testComputeIfAbsentConcurrently() throws Exception {
+        ToLongBiFunction<ICacheKey<String>, String> weigher = getMockWeigher();
```

Comment:
Nit: Should probably put comments on tests mentioning which test it's based on, if it's similar enough to an existing test